.github/workflows/ci.yml

```diff
@@ -416,3 +485,24 @@ jobs:
         run: nox -s build-emscripten
       - name: Test
         run: nox -s test-emscripten
+  
+  conclusion:
+    needs:
+      - fmt
+      - clippy
+      - check-target
+      - build
+      - valgrind
+      - careful
+      - coverage
+      - emscripten
+    if: always()
+    runs-on: ubuntu-latest
+    steps:
```

Comment:
Is this really necessary? Doesn't `needs` already require these job to finish successfully? Also what the reason to have `if: always()` here?

Generally speaking, I am not a fan of the manual JavaScript-based approach to CI workflows and would prefer if just split our workflows into one pull requests and for used by bors even if it means having some redundancies between the two workflow definitions if this allows us to keep the declarative approach.

I am also not sure if just adjusting the matrix of the `build` job is sufficient differentiation between the two use cases (pre-review check and pre-merge check). More fundamentally, I feel that the `build` could use some re-engineering as it has a collected a lot of different use cases in conditional steps which would work better as individual jobs IMHO.

Reply:
> Is this really necessary? Doesn't `needs` already require these job to finish successfully? Also what the reason to have `if: always()` here?

It gives a nicer overview: https://github.com/PyO3/pyo3/actions/runs/3508244920/jobs/5876898372#step:2:1 even when canceled due to previous failures.

Copied from https://github.com/cross-rs/cross/blob/ab99baf7f4c39403a0b033cef95d798bd0b9a940/.github/workflows/ci.yml#L361-L373

Reply:
> I feel that the `build` could use some re-engineering as it has a collected a lot of different use cases in conditional steps which would work better as individual jobs IMHO.

Expanding more jobs may result in more cache misses thus longer build time, we are [already over the 10GB cache limit](https://github.com/PyO3/pyo3/actions/caches).

Reply:
> Expanding more jobs may result in more cache misses thus longer build time, we are [already over the 10GB cache limit](https://github.com/PyO3/pyo3/actions/caches).

We could also improve on this front, considering the approach outlined in https://github.com/PyO3/pyo3/issues/2710#issuecomment-1296259994, the caches for the pull requests could become much more lightweight. (I think the issue of missing caches could also be solved by running the lightweight on-pull-request workflow as an on-push event on `main`? The heavyweight jobs would only run on `trying`/`staging` I guess?)

Reply:
> run check (not test) for each OS / Python version, and then one run of the test suite on each OS w. latest Python

It's not really lightweight IMO, I'd drop the `run check (not test) for each OS / Python version` part, and coverage jobs are even heavier but they should be kept to ensure good test coverage.

> I think the issue of missing caches could also be solved by running the lightweight on-pull-request workflow as an on-push event on `main`?

It will do, but I think with bors we shouldn't re-run workflow on `main`, [bors was designed to avoid broken checks on `main` I think?](https://graydon2.dreamwidth.org/1597.html). Maybe we should explore using [sccache](https://github.com/mozilla/sccache) instead.

Reply:
> I'd drop the run check (not test) for each OS / Python version part

I agree. I don't think full platform coverage is something we need during review.

> and coverage jobs are even heavier but they should be kept to ensure good test coverage.

As with the above, I think one coverage job should suffice. (And it could subsume the test job because the tests run to produce the coverage.)

> It will do, but I think with bors we shouldn't re-run workflow on main, 

Exactly, bors should only run on `trying/staging` which is `main` after merging the PR, hence one does not need to re-run bors on `main` because updating `main` to `trying/staging` is a fast-forward merge to something that was just tested.

So in summary:
* `main` and pull requests: lightweight pre-review workflow to seed cache.
* `trying/staging`: heavyweight bors pre-merge workflow. (Since bors pushes to our repo it should be able to use our caches without seeding.)